src/apps/app_bulk_upload.erl

```diff
@@ -0,0 +1,173 @@
+-module(app_bulk_upload).
+-export([upload/2, upload/3, download/2]).
+
+-include("ar.hrl").
+-include_lib("eunit/include/eunit.hrl").
+
+%%% Starts a server that accepts possibly huge files, splits them into small chunks, and
+%%% submits them to Arweave as separate transactions. Transactions are submitted via
+%%% the queue module.
+
+-define(MB, 1048576). % 1024 * 1024
+-define(BLOB_HASH_ALGO, sha256).
+
+%% @doc Starts a queue server, splits the given file into chunks, wraps the chunks as 
+%% Arweave transactions, and submits them to the queue. Returns the queue PID.
+upload(Wallet, Filename) when is_list(Filename) ->
+	{ok, Filecontents} = file:read_file(Filename),
+	upload(Wallet, Filecontents);
+upload(Wallet, Blob) when is_binary(Blob) ->
+	upload(whereis(http_entrypoint_node), Wallet, Blob).
+
+upload(Node, Wallet, Blob) ->
+	Queue = app_queue:start(Node, Wallet),
+	Hash = crypto:hash(?BLOB_HASH_ALGO, Blob),
+	upload_blob(Queue, Hash, Blob),
+	Queue.
+
+%% @doc Takes a binary blob and processes it chunk by chunk. Each chunk is converted into
+%% a transaction and put into the queue. Chunk size is 1MB.
+upload_blob(Queue, Hash, Blob) ->
+	BlobSize = byte_size(Blob),
+	ChunkNumber = BlobSize div ?MB + if BlobSize rem ?MB =/= 0 -> 1; true -> 0 end,
+	upload_blob(Queue, Hash, Blob, ChunkNumber, 1).
+
+upload_blob(Queue, Hash, Blob, ChunkNumber, ChunkPosition) ->
+	if byte_size(Blob) =< ?MB ->
+		app_queue:add(Queue, chunk_to_tx(Hash, Blob, ChunkNumber, ChunkPosition));
+	true ->
+		<< Chunk:?MB/binary, Rest/binary >> = Blob,
+		app_queue:add(Queue, chunk_to_tx(Hash, Chunk, ChunkNumber, ChunkPosition)),
+		upload_blob(Queue, Hash, Rest, ChunkNumber, ChunkPosition + 1)
+	end.
+
+%% @doc Converts the given binary chunk into a transaction. A hash of the whole block the chunk
+%% is part of is assigned as a tag.
+chunk_to_tx(Hash, Chunk, ChunkNumber, ChunkPosition) ->
+	#tx {
+		tags =
+			[
+				{"app_name", "BulkUpload"},
+				{"blob_hash", ar_util:encode(Hash)},
+				{"number_of_chunks", integer_to_binary(ChunkNumber)},
+				{"chunk_position", integer_to_binary(ChunkPosition)}
+			],
+		data = Chunk
+	}.
+
+%% @doc Searches the local storage for the chunks of the blob with the given hash.
+%% If the blob is reconstructed successfuly, writes it to the specified destination.
+%% The provided hash has to be a Base 64 encoded SHA 256 hash of the file as a binary string.
```

Comment:
I would then go a little further and accept a list of tags to assign to every tx here, not `Hash`. So I would pass the `blob_base64_sha256` tag along. If we want to attach other hashes in the future, we will add more tags.

Reply:
As for the variable names, yes - they do not have to carry all these details around.

As a tag name though, it may help people browsing the weave to make sense of transactions.